packages/client/src/knock.ts

```diff
@@ -74,11 +103,46 @@ class Knock {
 
   // Used to teardown any connected instances
   teardown() {
+    if (this.tokenExpirationTimer) {
+      clearTimeout(this.tokenExpirationTimer);
+    }
     if (!this.apiClient) return;
     if (this.apiClient.socket) {
       this.apiClient.socket.disconnect();
     }
   }
+
+  private async maybeScheduleUserTokenExpiration(
+    callbackFn: UserTokenExpiringCallback,
+    timeBeforeExpirationInMs: number = 1000,
+  ) {
+    if (!this.userToken) return;
+
+    const decoded = jwtDecode(this.userToken);
+    const expiresAtMs = (decoded.exp ?? 0) * 1000;
+    const nowMs = Date.now();
+
+    // Expiration is in the future
+    if (expiresAtMs && expiresAtMs > nowMs) {
+      // Check how long until the token should be regenerated
+      // | ----------------- | ----------------------- |
+      // ^ now               ^ expiration offset       ^ expires at
+      const msInFuture = expiresAtMs - timeBeforeExpirationInMs - nowMs;
+
+      this.tokenExpirationTimer = setTimeout(async () => {
+        const newToken = await callbackFn(this.userToken as string, decoded);
+        this.userToken = newToken;
+
+        // Resync socket connection
+        this.createApiClient();
```

Comment:
Do we need to teardown old connection before starting a new one here?

Reply:
Good question. I initially thought no since there was already a teardown/deconstructor for the api client, but yeah, maybe we should disconnect the socket to keep everything fresh.

Reply:
This does make this issue more apparent: https://github.com/knocklabs/javascript/issues/23

Toasts come through but the notification count doesn't update after the token changes

Reply:
Does this disconnect/reconnect solve the real-time issue? Do we need to do anything at the feed layer to reconnect the connected channels or does instantiating a new API client just handle that?

Reply:
It doesn't. Starting to investigate more and I think it might be an issue with how we memoize the knock and feed clients.